internal/fs/dir_handle.go

```diff
@@ -156,78 +174,66 @@ func fixConflictingNames(entries []fuseutil.Dirent) (err error) {
 	return
 }
 
-// Read all entries for the directory, fix up conflicting names, and fill in
-// offset fields.
-//
-// LOCKS_REQUIRED(in)
-func readAllEntries(
-	ctx context.Context,
-	in inode.DirInode) (entries []fuseutil.Dirent, err error) {
-	// Read one batch
-	var batch []fuseutil.Dirent
-	batch, ContinuationToken, err = in.ReadEntries(ctx, ContinuationToken)
-	if err != nil {
-		err = fmt.Errorf("ReadEntries: %w", err)
-		return
-	}
-	// Accumulate.
-	entries = append(entries, batch...)
-
-	// Ensure that the entries are sorted, for use in fixConflictingNames
-	// below.
-	// TODO: Fix this after  asynchronous fetch is added.
-	sort.Sort(sortedDirents(entries))
-
-	// Fix name conflicts.
-	err = fixConflictingNames(entries)
-	if err != nil {
-		err = fmt.Errorf("fixConflictingNames: %w", err)
-		return
-	}
-
-	// Return a bogus inode ID for each entry, but not the root inode ID.
-	//
-	// NOTE(jacobsa): As far as I can tell this is harmless. Minting and
-	// returning a real inode ID is difficult because fuse does not count
-	// readdir as an operation that increases the inode ID's lookup count and
-	// we therefore don't get a forget for it later, but we would like to not
-	// have to remember every inode ID that we've ever minted for readdir.
-	//
-	// If it turns out this is not harmless, we'll need to switch to something
-	// like inode IDs based on (object name, generation) hashes. But then what
-	// about the birthday problem? And more importantly, what about our
-	// semantic of not minting a new inode ID when the generation changes due
-	// to a local action?
-	for i, _ := range entries {
-		entries[i].Inode = fuseops.RootInodeID + 1
-	}
-
-	return
-}
+// Fetch Dirent entries from GCSfuse.Will be used as a goroutine which is run asynchronously
+// to fetch data in the background while kernel requests are also served simultaneously.
+func (dh *DirHandle) FetchEntriesAsync(
+	rootInodeId int,
+	firstCall bool) {
+	ctx, cancel := context.WithCancel(context.TODO())
+	var err error
+	for ContinuationToken != "" || firstCall {
+		var entries []fuseutil.Dirent
+		dh.In.Lock()
+		entries, ContinuationToken, err = dh.In.ReadEntries(ctx, ContinuationToken)
+		dh.In.Unlock()
+		if err != nil {
+			err = fmt.Errorf("ReadEntries: %w", err)
+			rec.Lock()
+			rec.err = err
+			rec.Unlock()
+			rec.cond.Broadcast()
+			cancel()
```

Comment:
. We are waiting for the gcs call to complete and then doing the processing, so why do you need to cancel the context here.

Reply:
Since this was also an error situation, the goroutine must be stopped.It was advised to cancel the context to release the resources associated with it.

Reply:
the network call is finished by the time you come here. what else the cancel will do? what resources are held by the context?

Reply:
I read up on it.Since the network call is finished ,there is no need to explicitly cancel the context as it is implicitly cleaned up by go runtime.Will update the code.